src/cascadia/TerminalControl/XamlLights.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "TermControl.h"
+#include "XamlLights.h"
+#include "VisualBellLight.g.cpp"
+
+using namespace winrt::Windows::UI::Xaml;
+using namespace winrt::Windows::UI::Xaml::Media;
+
+namespace winrt::Microsoft::Terminal::Control::implementation
+{
+    DependencyProperty VisualBellLight::_IsTargetProperty{ nullptr };
+
+    VisualBellLight::VisualBellLight()
+    {
+        _InitializeProperties();
+    }
+
+    void VisualBellLight::_InitializeProperties()
+    {
+        // Initialize any dependency properties here.
+        // This performs a lazy load on these properties, instead of
+        // initializing them when the DLL loads.
+        if (!_IsTargetProperty)
+        {
+            _IsTargetProperty =
+                DependencyProperty::RegisterAttached(
+                    L"IsTarget",
+                    winrt::xaml_typename<bool>(),
+                    winrt::xaml_typename<Control::VisualBellLight>(),
+                    PropertyMetadata{ winrt::box_value(false), PropertyChangedCallback{ &VisualBellLight::OnIsTargetChanged } });
+        }
+    }
+
+    // Method Description:
+    // - This function is called when the first target UIElement is shown on the screen,
+    //   this enables delaying composition object creation until it's actually necessary.
+    // Arguments:
+    // - newElement: unused
+    void VisualBellLight::OnConnected(UIElement const& /* newElement */)
+    {
+        if (!CompositionLight())
+        {
+            auto spotLight{ Window::Current().Compositor().CreateAmbientLight() };
+            spotLight.Color(Windows::UI::Colors::White());
+            CompositionLight(spotLight);
+        }
+    }
+
+    // Method Description:
+    // - This function is called when there are no more target UIElements on the screen
+    // - Disposes of composition resources when no longer in use
+    // Arguments:
+    // - oldElement: unused
+    void VisualBellLight::OnDisconnected(UIElement const& /* oldElement */)
+    {
+        if (CompositionLight())
+        {
+            CompositionLight(nullptr);
+        }
+    }
+
+    winrt::hstring VisualBellLight::GetId()
+    {
+        return VisualBellLight::GetIdStatic();
+    }
+
+    void VisualBellLight::OnIsTargetChanged(DependencyObject const& d, DependencyPropertyChangedEventArgs const& e)
+    {
+        const auto& uielem{ d.try_as<UIElement>() };
+        const auto& brush{ d.try_as<Brush>() };

```

Comment:
`try_as` doesn't returns `winrt::com_ptr` by value and this code only compiles, because you bind it as a const reference (since it's not possible to create references to temporary values normally).
The compiler puts the return value of `try_as` transparently on the stack and a reference into the `uielem`. In Release builds this is "imperfection" certainly optimized away, but it'd be better and more correct to write this instead:
```cpp
const auto uielem = d.try_as<UIElement>();
const auto brush = d.try_as<Brush>();
```